apps/docs/app/ui/components/componentExample/ComponentExample.tsx

```diff
@@ -0,0 +1,101 @@
+"use client";
+
+import { useState, memo, useCallback, useMemo, type ReactNode } from "react";
+import clsx from "clsx";
+
+import { CodeIcon } from "@/components/icon";
+import { ToggleButton } from "@/components/toggleButton/ToggleButton.tsx";
+
+import ComponentPreviewWrapper from "./ComponentPreviewWrapper.tsx";
+
+import "./componentExample.css";
+
+interface CommonProps {
+    className?: string;
+    isOpen?: boolean;
+}
+
+interface CodeProps extends CommonProps {
+    type: "code";
+    code: ReactNode;
+}
+
+interface PreviewProps extends CommonProps {
+    type: "preview";
+    preview: ReactNode;
+}
+
+interface BothProps extends CommonProps {
+    type: "both";
+    code: ReactNode;
+    preview: ReactNode;
+}
+
+export type ComponentExampleProps = CodeProps | PreviewProps | BothProps;
+
+const ComponentExample = memo(({
+    type = "both",
+    className,
+    isOpen = false,
+    ...props
+}: ComponentExampleProps) => {
+    const [showCode, setShowCode] = useState(isOpen);
+
+    const showBothComponent = useMemo(() => type === "both", [type]);
+    const showPreviewComponent = useMemo(() => type === "preview" || showBothComponent, [type, showBothComponent]);
+    const showCodeComponent = useMemo(() => (showBothComponent && showCode) || type === "code", [showBothComponent, showCode, type]);
+
+    const toggleShowCode = useCallback(() => setShowCode(prevShowCode => !prevShowCode), []);
```

Comment:
just a FYI, if you end up using this pattern frequently, i've seen and use a "use-toggle" hook a couple of time. It's a really simple reusable hook

```
import { useCallback, useState } from 'react'

import type { Dispatch, SetStateAction } from 'react'

export function useToggle(
  defaultValue?: boolean,
): [boolean, () => void, Dispatch<SetStateAction<boolean>>] {
  const [value, setValue] = useState(!!defaultValue)

  const toggle = useCallback(() => {
    setValue(x => !x)
  }, [])

  return [value, toggle, setValue]
}

```